From 36f74f7e35682081d561d9b290413cca7b443d7d Mon Sep 17 00:00:00 2001
From: Adriana Kobylak <anoo@us.ibm.com>
Date: Fri, 17 Jul 2020 15:36:22 -0500
Subject: [PATCH] volatile: Preserve original in eMMC

The eMMC only has one directory. As workaround, preserve the
original VOLATILE partition since HB expects the BMC to clear
it and restore the original.

Change-Id: I924fb1c2cbf3c641add4325689be75361a449dce
Signed-off-by: Adriana Kobylak <anoo@us.ibm.com>
---
 mmc/obmc-flash-bios   | 6 ++++++
 vpnor/obmc-vpnor-util | 8 ++++++++
 2 files changed, 14 insertions(+)

diff --git a/mmc/obmc-flash-bios b/mmc/obmc-flash-bios
index 0a496ec..623d0d1 100644
--- a/mmc/obmc-flash-bios
+++ b/mmc/obmc-flash-bios
@@ -39,6 +39,12 @@ mmc_init() {
     mount ${base_dir}/hostfw-${boot_label} ${tmp_dir} -o ro
     rm -f ${running_dir}/*
     cp -p ${tmp_dir}/* ${running_dir}/
+
+    # Workaround. Copy off the original HB_VOLATILE partition since HB expects
+    # the BMC to clear it and restore it to default.
+    mkdir -p "${base_dir}/volatile"
+    cp ${tmp_dir}/*VOLATILE* ${base_dir}/volatile/ || true
+
     umount ${tmp_dir}
     rm -rf ${tmp_dir}
 
diff --git a/vpnor/obmc-vpnor-util b/vpnor/obmc-vpnor-util
index b529f47..eb63f9b 100644
--- a/vpnor/obmc-vpnor-util
+++ b/vpnor/obmc-vpnor-util
@@ -23,11 +23,19 @@ clear_volatile() {
     if [ -f "${rwVolatile}" ]; then
       echo "Clear $rwVolatile"
       rm "${rwVolatile}"
+      mmcVolatile="/media/hostfw/volatile/"
+      if [ -d "${mmcVolatile}" ]; then
+        cp ${mmcVolatile}/* "${rwVolatile}"
+      fi
     fi
     prsvVolatile="${PNOR_PRSV_ACTIVE_PATH}${volatileName}"
     if [ -f "${prsvVolatile}" ]; then
       echo "Clear $prsvVolatile"
       rm "${prsvVolatile}"
+      mmcVolatile="/media/hostfw/volatile/"
+      if [ -d "${mmcVolatile}" ]; then
+        cp ${mmcVolatile}/* "${prsvVolatile}"
+      fi
     fi
   done
   # Always reset the sensor after clearing
-- 
1.8.3.1

